torchtext/csrc/clip_tokenizer.cpp

```diff
@@ -0,0 +1,137 @@
+#include <clip_tokenizer.h>
+#include <regex.h>  // @manual
+
+#include <unordered_set>
+
+namespace torchtext {
+const Regex kCLIPRegex(
+    "(?i)(<\\|startoftext\\|>|<\\|endoftext\\|>|\\'s|\\'t|\\'re|\\'ve|"
+    "\\'m|\\'ll|\\'d|[\\pL]+|[\\pN]|[^\\s\\pL\\pN]+)");
+const std::string kWhitespaceString("</w>");
+const std::unordered_set<std::string> kSpecialTokens{"<|startoftext|>",
+                                                     "<|endoftext|>"};
+
+std::vector<std::string> clip_pre_tokenizer(std::string input) {
+  std::string token;
+  std::vector<std::string> tokens;
+  re2::StringPiece inp(input);
+  while (kCLIPRegex.FindAndConsume(&inp, &token)) {
+    tokens.push_back(token);
+  }
+  return tokens;
+}
+
+std::vector<std::string> CLIPEncoder::BPE_(
+    const std::vector<std::string> &token_list) {
+  // Given a list of input tokens, keep finding the best bpe merge and
+  // generate a new list of tokens until
+  //  1) token list size reduced to 1
+  //      OR
+  //  2) can't find bpe merge
+  auto concatenated = concatenate_strings(token_list);
+  if (caching_enabled_ && cache_.contains(concatenated)) {
+    return cache_.at(concatenated);
+  } else if (kSpecialTokens.find(concatenated) != kSpecialTokens.end()) {
+    return {concatenated};
+  }
+
+  std::vector<std::string> tok_list(token_list.begin(), token_list.end() - 1);
+  tok_list.push_back(token_list[token_list.size() - 1] + kWhitespaceString);
+  auto pairs = get_pairs(tok_list, seperator_);
+  if (pairs.empty()) {
+    return {concatenated + kWhitespaceString};
+  }
+  while (true) {
+    auto bigram = FindBestPair_(pairs);
+    if (!bpe_merge_ranks_.contains(bigram)) break;
+
+    // Finding all indexes that token_list[i] == first and token_list[i+1] ==
+    // second. After the loop, new token list will be
+    //  1) first + second pair
+    //  2) all the other tokens in the original token list
+    //
+    // For example: first="a" second="w" and token_list =
+    // ["a", "w", "some", "a", "w", "e"]
+    // Result: new_token_list = ["aw", "some", "aw", "e"]
+
+    auto parts = split_tokens(bigram, seperator_);
+    std::vector<std::string> new_token_list;
+    std::size_t i = 0;
+    while (i < tok_list.size()) {
+      auto j = list_str_index(tok_list, parts.first, i);
+      if (j != -1) {
+        for (int k = i; k < j; k++) new_token_list.push_back(tok_list[k]);
```

Comment:
Split into multiple lines?

Reply:
Thanks for sharing this. Currently this file is formatted with Google C++ Style . We are in the process of setting up clang-format for torchtext, so all formatting can be handled there.